### PR TITLE
CBMC proofs for Device Defender Client library.

### PR DESCRIPTION
**Pull request by aggarg**
Description
-------------

CBMC proofs for Device Defender Client library.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by sarenameas — DISMISSED** [review]

File: source/defender.c [hunk 1]
```diff
@@ -265,18 +266,19 @@ static void writeFormatAndSuffix( char * pBuffer,
                     DEFENDER_API_ACCEPTED_SUFFIX_LENGTH );
             break;
 
+        /* The default is here just to silence compiler warnings in a way which
+         * does not bring coverage down. The assert at the beginning of this
+         * function ensures that the only api value hitting this case can be
+         * DefenderCborReportRejected. */
```
Comment (karkhaz) [inline source/defender.c:272]: @markrtuttle I wonder if there's a way to enforce that this comment remains true in CBMC?

For example, could we move the code up to the `case` block and put `assert(0)` in the `default` block? Or one of CBMC's macros like `__CPROVER_assert(0)`?

File: test/cbmc/.gitignore [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# Emitted by CBMC Viewer.
+TAGS-*
+
+# Emitted by Arpa
+arpa_cmake/
+arpa-validation-logs/
+Makefile.arpa
+
+# These files should be overwritten whenever prepare.py runs.
+cbmc-batch.yaml
```
Comment (karkhaz) [inline test/cbmc/.gitignore:16]: No need for this line I think. We're not generating or committing `cbmc-batch.yaml` any more.

**Review by karkhaz — CHANGES_REQUESTED** [review]
In order for this to work, you'll also need to add a 'run script' to this PR. [Here is the run script](https://github.com/FreeRTOS/coreHTTP/blob/master/test/cbmc/proofs/run-cbmc-proofs.py) from coreHTTP as an example. You can almost certainly just use exactly the same file, but just replace "coreHTTP" in the file with Defender.

You will also need to add a 'marker file' to every proof directory, to indicate that it contains a proof. The coreHTTP repository has a text file called [`cbmc-proof.txt`](https://github.com/FreeRTOS/coreHTTP/blob/master/test/cbmc/proofs/run-cbmc-proofs.py#L26) in every proof directory; it's just a [`text file with some standard text`](https://github.com/FreeRTOS/coreHTTP/blob/master/test/cbmc/proofs/httpParserOnStatusCallback/cbmc-proof.txt). I see that you've written some nice READMEs for every proof, so you might want to rename all the READMEs to cbmc-proof.txt.

You can test all of this out locally even before we deploy CI: just run the run script, it should run all the proofs in parallel.

File: test/cbmc/.gitignore (hunk 2 above)
Comment (markrtuttle) [reply test/cbmc/.gitignore:16]: But we do need some touch file like cbmc-batch.yaml in the proof directories to mark the directory as containing a proof.  I don't see any in this pull request.

**Comment by markrtuttle** [comment]
Do not follow the advice to rename the README files cbmc-proof.txt.   The cbmc-proof.txt is just supposed to be a touch file, its contents should not have any value.

File: source/defender.c (hunk 1 above)
Comment (markrtuttle) [reply source/defender.c:272]: I think adding a
```
    default:
        assert(0);
```
is a reasonable idea.

**Review by markrtuttle — DISMISSED** [review]

**Review by sarenameas — DISMISSED** [review]

File: test/cbmc/proofs/Defender_MatchTopic/Defender_MatchTopic_harness.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
```
Comment (karkhaz) [inline test/cbmc/proofs/Defender_MatchTopic/Defender_MatchTopic_harness.c:2]: I think this should be
```
Copyright Amazon.com Inc. or its affiliates.
```
on all files

File: test/cbmc/proofs/run-cbmc-proofs.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+This tool automates the process of running `make report` in each of
+the CBMC proof directories. The tool calculates the dependency graph
+of all tasks needed to build, run, and report on all the proofs, and
+executes these tasks in parallel.
+
+The tool is roughly equivalent to doing this:
+
+        litani init --project "FreeRTOS Device Defender";
+
+        for proof in $(find . -name wellspring.txt); do
```
Comment (karkhaz) [inline test/cbmc/proofs/run-cbmc-proofs.py:41]: Please change to
```
for proof in $(find . -name cbmc-proof.txt); do
```

**Review by karkhaz — CHANGES_REQUESTED** [review]

**Review by karkhaz — APPROVED** [review]

**Review by sarenameas — APPROVED** [review]

**Pull request merged**